src/mobu/business/notebookrunner.py

```diff
@@ -0,0 +1,178 @@
+        self.logger.info("Repository cloned and ready")
+        await super().startup()
+        self._last_login = self._now()
+        await self.initial_delete_lab()
+
+    def clone_repo(self) -> None:
+        url = self.options.get("repo_url", REPO_URL)
+        branch = self.options.get("repo_branch", REPO_BRANCH)
+        path = self._repo_dir.name
+        with self.timings.start("clone_repo"):
+            self._repo = git.Repo.clone_from(url, path, branch=branch)
+
+    async def initial_delete_lab(self) -> None:
+        with self.timings.start("initial_delete_lab"):
+            await self._client.delete_lab()
+
+    async def lab_business(self) -> None:
+        self._next_notebook()
+        assert self.notebook
+
+        await self.ensure_lab()
+        await self.lab_settle()
+        kernel = await self.create_kernel()
+
+        self.logger.info(f"Starting notebook: {self.notebook.name}")
+        cells = self.read_notebook(self.notebook.name, self.notebook.path)
+        nb_iterations = self.options.get("notebook_iterations", 1)
+        for count in range(nb_iterations):
+            iteration = f"{count + 1}/{nb_iterations}"
+            msg = f"Notebook '{self.notebook.name}' iteration {iteration}"
+            self.logger.info(msg)
+
+            await self._reauth_if_needed()
+
+            for cell in cells:
+                self.code = "".join(cell["source"])
+                await self.execute_code(kernel, self.code)
+
+        await self.delete_kernel(kernel)
+        self.logger.info(f"Success running notebook: {self.notebook.name}")
+
+    async def lab_settle(self) -> None:
+        with self.timings.start("lab_settle"):
+            await asyncio.sleep(self.options.get("settle_time", 0))
+
+    def read_notebook(self, name: str, path: str) -> List[Dict[str, Any]]:
+        with self.timings.start(f"read_notebook:{name}"):
+            notebook_text = Path(path).read_text()
+            cells = json.loads(notebook_text)["cells"]
+        return [c for c in cells if c["cell_type"] == "code"]
+
+    async def create_kernel(self) -> str:
+        self.logger.info("create_kernel")
+        with self.timings.start("create_kernel"):
+            kernel = await self._client.create_kernel()
+        return kernel
+
+    async def execute_code(self, kernel: str, code: str) -> None:
+        self.logger.info("Executing:\n%s\n", code)
+        with self.timings.start("run_code", {"code": code}) as sw:
+            reply = await self._client.run_python(kernel, code)
+            sw.annotation["result"] = reply
+        self.logger.info(f"Result:\n{reply}\n")
+
+    async def delete_kernel(self, kernel: str) -> None:
+        self.logger.info(f"Deleting kernel {kernel}")
+        with self.timings.start("delete_kernel"):
+            await self._client.delete_kernel(kernel)
+
+    def dump(self) -> dict:
+        r = super().dump()
+        r["running_code"] = self.code
```

Comment:
Good catch. I think it should only dump the running code if in fact it is currently running that code (and it's taking a really long time or something), so rather than renaming it, I am changing the code to clear that variable when the code finishes running.